File: CAN_DRIVER/NANOTEC/NanoLib_1.0.1/NanoLib C# Windows/nanolib_csharp_win_1.0.1/Example/NanolibExample/NanolibSamplerExample.cs
using NanolibCsharpExampleWin;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace NanolibCsharpExampleWin
{
    internal delegate void ProcessSampledData(Nlc.SampleDataVector sampleDatas);

    /// <summary>
    /// SamplerNotifyCallback class derived from Nlc.SamplerNotify.
    /// </summary>
    internal class SamplerNotifyCallback : Nlc.SamplerNotify
    {
        private readonly EventWaitHandle waitHandle = new EventWaitHandle(false, EventResetMode.AutoReset);
        readonly ProcessSampledData processSampleData;

        /// <summary>
        /// Creates sampler notification callback.
        /// </summary>
        /// <param name="processSampleData">Sampled data processing delegate.</param>
        public SamplerNotifyCallback(ProcessSampledData processSampleData) : base()
        {
            this.processSampleData = processSampleData;
        }

        /// <summary>
        /// Callback used for notification on sampler events
        /// </summary>
        /// <param name="lastError">Last sampler error</param>
        /// <param name="samplerState">Current sampler state</param>
        /// <param name="sampleDatas">New sampler data</param>
        /// <param name="applicationData">Application supplied data (see Sampler.start(..))</param>
        public override void notify(Nlc.ResultVoid lastError, Nlc.SamplerState samplerState, Nlc.SampleDataVector sampleDatas, long applicationData)
        {
            // Be aware that notifications are executed in the context of separate threads 
            // other than thread that started the sampler.
            // 
            // Be careful when calling Nanolib functionality here, as doing so may cause this method
            // to be called recursively, potentially causing your application to deadlock.
            // 
            // For the same reason, this method should not throw exceptions.

            processSampleData(sampleDatas);

            if ((samplerState != Nlc.SamplerState.Ready) && (samplerState != Nlc.SamplerState.Running))
            {
                if (samplerState == Nlc.SamplerState.Failed)
                {
                    Console.Error.WriteLine("Sampler execution failed with error: {0}", lastError.getError());
                }

                // It's now safe to destroy this notification object
                waitHandle.Set();
            }
        }

        /// <summary>
        /// Wait for the sampler to finish.
        /// </summary>
        public void Wait()
        {
            waitHandle.WaitOne();
        }

        /// <summary>
        /// Wait for the sampler to finish.
        /// </summary>
        /// <param name="timeOut">Timeout value.</param>
        /// <returns>True if sampler has finished, false on timeout.</returns>
        public bool Wait(int timeOut)
        {
            return waitHandle.WaitOne(timeOut);
        }
    }

    /// <summary>
    /// Examples of using NanoLib sampler.
    /// </summary>
    internal class NanolibSamplerExample
    {
        private readonly Nlc.SamplerInterface sampler;
        private readonly NanolibHelper helper;
        private readonly Nlc.DeviceHandle deviceHandle;
        private readonly Nlc.OdIndex triggerAddress = new Nlc.OdIndex(0x2400, 1); // We just use first NanoJ input for the example
        private readonly Nlc.OdIndexVector trackedAddresses = new Nlc.OdIndexVector(
                new Nlc.OdIndex[] {
                    new Nlc.OdIndex(0x230F, 0), // Up time in seconds
                    new Nlc.OdIndex(0x4014, 3) // Temperature
                });
        private readonly string[] addressNames = new string[] { "Up time", "Temperature" };

        private readonly Nlc.SamplerConfiguration configuration = new Nlc.SamplerConfiguration();

        private bool headerPrinted;

        ulong sampleNumber;
        ulong lastIteration = ulong.MaxValue;

        /// <summary>
        /// Creates sampler example.
        /// </summary>
        /// <param name="helper">NanolibHelper instance.</param>
        /// <param name="deviceHandle">Nlc.DeviceHandle on which sampler will be executed.</param>
        public NanolibSamplerExample(NanolibHelper helper, Nlc.DeviceHandle deviceHandle)
        {            
            this.helper = helper;
            sampler = helper.GetSamplerInterface();
            this.deviceHandle = deviceHandle;
        }

        /// <summary>
        /// Run sampler examples.
        /// </summary>
        public void RunSamplerExamples()
        {
            SamplerExamplerWithoutNotification();
            SamplerExamplesWithNotification();
        }

        private void HandleSamplerFailed()
        {
            Nlc.ResultVoid lastError = sampler.getLastError(deviceHandle);

            Console.Error.WriteLine("Sampler execution failed with error: {0}", lastError.getError());
        }

        /// <summary>
        /// Examples without notification(callback).
        /// </summary>
        private void SamplerExamplerWithoutNotification()
        {
            ProcessSamplerWithoutNotificationNormal();
            ProcessSamplerWithoutNotificationRepetative();
            ProcessSamplerWithoutNotificationContinuous();
        }

        /// <summary>
        /// Examples with notification(callback).
        /// </summary>
        private void SamplerExamplesWithNotification()
        {
            ProcessSamplerWithNotificationNormal();
            ProcessSamplerWithNotificationRepetative();
            ProcessSamplerWithNotificationContinuous();
        }

        /// <summary>
        /// Sampler example in normal mode without notification.
        /// </summary>
        private void ProcessSamplerWithoutNotificationNormal()
        {
            Console.WriteLine("Sampler without notification in normal mode:");

            SetSamplerConfiguration(Nlc.SamplerMode.Normal);

            StartSampler(null);

            Nlc.SamplerState state;

            do
            {
                Thread.Sleep(100);
                ProcessSampledData(GetSampledData());
                state = GetSamplerState();
            } while (Nlc.SamplerState.Ready == state || Nlc.SamplerState.Running == state);

            ProcessSampledData(GetSampledData()); // Get last remaining data
        }

        /// <summary>
        /// Sampler example in repetative mode without notification.
        /// </summary>
        private void ProcessSamplerWithoutNotificationRepetative()
        {
            Console.WriteLine("Sampler without notification in repetative mode:");

            SetSamplerConfiguration(Nlc.SamplerMode.Repetitive);

            StartSampler(null);

            Nlc.SamplerState state = Nlc.SamplerState.Ready;

            do
            {
                Thread.Sleep(100);
                ProcessSampledData(GetSampledData());

                if (lastIteration >= 4)
                {
                    // In repeat mode the sampler will continue to run until it is stopped or an error occurs
                    sampler.stop(deviceHandle);
                    break;
                }

                state = GetSamplerState();
            } while (Nlc.SamplerState.Ready == state || Nlc.SamplerState.Running == state);

            ProcessSampledData(GetSampledData()); // Get last remaining data

            if (state == Nlc.SamplerState.Failed)
                HandleSamplerFailed();
        }

        /// <summary>
        /// Sample example in continuous mode without notification.
        /// </summary>
        private void ProcessSamplerWithoutNotificationContinuous()
        {
            Console.WriteLine("Sampler without notification in continuous mode:");
            SetSamplerConfiguration(Nlc.SamplerMode.Continuous);

            StartSampler(null);

            Nlc.SamplerState state = Nlc.SamplerState.Ready;

            int seconds = 0;

            do
            {
                Thread.Sleep(1000);
                seconds++;
                ProcessSampledData(GetSampledData());

                if (seconds >= 10)
                {
                    // In continuous mode the sampler will continue to run until it is stopped or an error occurs
                    sampler.stop(deviceHandle);
                    break;
                }

                state = GetSamplerState();
            } while (Nlc.SamplerState.Ready == state || Nlc.SamplerState.Running == state);

            ProcessSampledData(GetSampledData()); // Get last remaining data

            if (state == Nlc.SamplerState.Failed)
                HandleSamplerFailed();
        }

        /// <summary>
        /// Sampler example in normal node with notification.
        /// </summary>
        private void ProcessSamplerWithNotificationNormal()
        {
            Console.WriteLine("Sampler with notification in normal mode:");

            SetSamplerConfiguration(Nlc.SamplerMode.Normal);

            // Create sampler notification callback
            SamplerNotifyCallback callback = new SamplerNotifyCallback(ProcessSampledData);

            StartSampler(callback);

            callback.Wait();
        }

        /// <summary>
        /// Sampler example in repetative mode with notification.
        /// </summary>
        private void ProcessSamplerWithNotificationRepetative()
        {
            Console.WriteLine("Sampler with notification in repetative mode:");

            SetSamplerConfiguration(Nlc.SamplerMode.Repetitive);

            // Create sampler notification callback
            SamplerNotifyCallback callback = new SamplerNotifyCallback(ProcessSampledData);

            StartSampler(callback);

            while (!callback.Wait(100))
            {
                if (lastIteration >= 4)
                {
                    // In repeat mode the sampler will continue to run until it is stopped or an error occurs
                    sampler.stop(deviceHandle);
                }
            }
        }

        /// <summary>
        /// Sampler example in cotinuous mode with notification.
        /// </summary>
        private void ProcessSamplerWithNotificationContinuous()
        {
            Console.WriteLine("Sampler with notification in continuous mode:");
            SetSamplerConfiguration(Nlc.SamplerMode.Continuous);

            // Create sampler notification callback
            SamplerNotifyCallback callback = new SamplerNotifyCallback(ProcessSampledData);

            StartSampler(callback);

            Thread.Sleep(10000);
            sampler.stop(deviceHandle); // In continuous the sampler will continue to run until it is stopped or an error occurs
            // It's now safe to destroy the notification object(SamplerNotifyCallback callback)
        }

        /// <summary>
        /// Set sampler configuration.
        /// </summary>
        /// <param name="mode">Sampler mode.</param>
        private void SetSamplerConfiguration(Nlc.SamplerMode mode)
        {
            configuration.numberOfSamples = 5;
            configuration.periodMilliseconds = 1000;
            configuration.preTriggerNumberOfSamples = 0; // Unused currently
            configuration.trackedAddresses = trackedAddresses;
            configuration.triggerAddress = triggerAddress;
            configuration.triggerCondition = Nlc.SamplerTriggerCondition.TC_GREATER;
            configuration.triggerValue = 10;
            configuration.forceSoftwareImplementation = Nlc.SamplerMode.Continuous == mode;            
            configuration.mode = mode;

            Nlc.ResultVoid result = sampler.configure(deviceHandle, configuration);
            if (result.hasError())
            {
                throw new NanolibException("Unable to configure the sampler: " + result.getError());
            }
        }

        /// <summary>
        /// Starts the sampler.
        /// </summary>
        /// <param name="notify"></param>
        private void StartSampler(Nlc.SamplerNotify notify, Int64 applicationData = 0)
        {
            headerPrinted = false;
            sampleNumber = 0;
            lastIteration = ulong.MaxValue;

            if (trackedAddresses.Count < 1 || trackedAddresses.Count > 12)
            {
                throw new InvalidOperationException("The number of tracked addresses must be between 1 and 12");
            }

            if (trackedAddresses.Count != addressNames.Length)
            {
                throw new InvalidOperationException("The number of tracked addresses is not equal to the number of address names");
            }

            helper.WriteNumber(deviceHandle, 0, triggerAddress, 32); // Reset the trigger

            Nlc.ResultVoid result = sampler.start(deviceHandle, notify, applicationData);
            if (result.hasError())
            {
                throw new NanolibException("Unable to start the sampler: " + result.getError());
            }

            helper.WriteNumber(deviceHandle, 11, triggerAddress, 32); // Fire the trigger
        }

        /// <summary>
        /// Process the sampled data.
        /// </summary>
        /// <param name="sampleDatas">Sampled data</param>
        private void ProcessSampledData(Nlc.SampleDataVector sampleDatas)
        {
            if (sampleDatas?.Count > 0)
            {
                foreach (var sampleData in sampleDatas)
                {
                    if (sampleData.sampledValues.Count % addressNames.Length != 0)
                    {
                        throw new InvalidOperationException("SamplerNotifyCallback is not configured properly");
                    }

                    if (lastIteration != sampleData.iterationNumber)
                    {
                        lastIteration = sampleData.iterationNumber;
                        sampleNumber = 0;
                    }

                    if (!headerPrinted)
                    {
                        Console.WriteLine("------------------------------------------------------------");
                        Console.Write("{0,-11}{1,-9}", "Interation", "Sample");
                        foreach (var addressName in addressNames)
                        {
                            Console.Write("{0,-14}{1,-8}", string.Format("[{0}]", addressName), "Time");
                        }
                        Console.WriteLine();
                        Console.WriteLine("------------------------------------------------------------");
                        headerPrinted = true;
                    }

                    for (int i = 0; i < sampleData.sampledValues.Count; i += addressNames.Length)
                    {
                        StringBuilder output = new StringBuilder();
                        output.Append(String.Format("{0,-11}{1,-9}", sampleData.iterationNumber, sampleNumber++));
                        for (int j = 0; j < addressNames.Length; j++)
                        {
                            output.Append(String.Format("{0,-14}{1,-8}", sampleData.sampledValues[i + j].value, sampleData.sampledValues[i + j].collectTimeMsec));
                        }

                        Console.WriteLine(output.ToString());
                    }
                }
            }
        }

        /// <summary>
        /// Gets actual sampler state.
        /// </summary>
        /// <returns></returns>
        private Nlc.SamplerState GetSamplerState()
        {
            Nlc.ResultSamplerState resultState = sampler.getState(deviceHandle);

            if (resultState.hasError())
            {
                throw new NanolibException("Unable to get sampler state: " + resultState.getError());
            }

            return resultState.getResult();
        }

        /// <summary>
        /// Gets buffered sampled data.
        /// </summary>
        /// <returns></returns>
        private Nlc.SampleDataVector GetSampledData()
        {
            Nlc.ResultSampleDataArray resultData = sampler.getData(deviceHandle);

            if (resultData.hasError())
            {
                throw new NanolibException("Unable to get sampler data: " + resultData.getError());
            }

            return resultData.getResult();
        }
    }
}

File: CAN_DRIVER/NANOTEC/NanoLib_1.0.1/NanoLib C++ Linux/nanolib_cpp_linux_1.0.1/Example/src/nanolib_sampler_example.cpp
#include "nanolib_sampler_example.hpp"

#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <thread>

const NanoLibSamplerExample::TrackedAddress NanoLibSamplerExample::trackedAddresses[]
	= {{"Up time", nlc::OdIndex(0x230F, 0x00)}, {"Temperature", nlc::OdIndex(0x4014, 0x03)}};
// In this sample we use first NanoJ input as trigger
const nlc::OdIndex NanoLibSamplerExample::triggerAddress = nlc::OdIndex(0x2400, 0x01);
const nlc::SamplerTriggerCondition NanoLibSamplerExample::triggerCondition
	= nlc::SamplerTriggerCondition::TC_GREATER;
const uint32_t NanoLibSamplerExample::triggerValue = 10;
// Depends from condition
const uint32_t NanoLibSamplerExample::triggerValueInactive = NanoLibSamplerExample::triggerValue;
// Depends from condition
const uint32_t NanoLibSamplerExample::triggerValueActive = NanoLibSamplerExample::triggerValue + 1;
const uint16_t NanoLibSamplerExample::periodMilliseconds = 1000;
const uint16_t NanoLibSamplerExample::numberOfSamples = 5;


NanoLibSamplerExample::NanoLibSamplerExample(NanoLibHelper &helper,
											   const nlc::DeviceHandle connectedDeviceHandle)
	: nanolibHelper(helper), deviceHandle(connectedDeviceHandle) {
}

NanoLibSamplerExample::~NanoLibSamplerExample() {
}

NanoLibSamplerExample::SamplerNotifyExample::SamplerNotifyExample(
	NanoLibSamplerExample &example)
	: samplerExample(example), samplerRunning(true) {
}

NanoLibSamplerExample::SamplerNotifyExample::~SamplerNotifyExample() {
	// Destroying this notification object is safe only when the sampler is not active
	assert(!samplerRunning);
}

void NanoLibSamplerExample::SamplerNotifyExample::notify(
	const nlc::ResultVoid &lastError, const nlc::SamplerState samplerState,
	const std::vector<nlc::SampleData> &sampleDatas, int64_t applicationData) {

	// Be aware that notifications are executed in the context of separate threads 
	// other than thread that started the sampler.
	// 
	// Be careful when calling Nanolib functionality here, as doing so may cause this method
	// to be called recursively, potentially causing your application to deadlock.
	// 
	// For the same reason, this method should not throw exceptions.

	assert(samplerRunning);

	(void)applicationData;

	if (!sampleDatas.empty())
		samplerExample.processSampledData(sampleDatas);

	if (samplerState == nlc::SamplerState::Failed) {
		try {
			samplerExample.handleSamplerFailed(&lastError);
		} catch (...) {
			// see comment above
		}
	}

	if ((samplerState != nlc::SamplerState::Ready)
		&& (samplerState != nlc::SamplerState::Running)) {
		// It's now safe to destroy this notification object
		samplerRunning = false;
	}
}

void NanoLibSamplerExample::process() {

	processExamplesWithoutNotification();
	processExamplesWithNotification();
}

void NanoLibSamplerExample::processExamplesWithoutNotification() {

	processSamplerWithoutNotificationNormal();
	processSamplerWithoutNotificationRepetitive();
	processSamplerWithoutNotificationContinuous();
}

void NanoLibSamplerExample::processSamplerWithoutNotificationNormal() {

	const std::chrono::milliseconds sleepTimeMsec(periodMilliseconds);

	std::cout << "\nSampler without notification in normal mode:" << std::endl;

	configure(nlc::SamplerMode::Normal);
	start();

	nlc::SamplerState samplerState;

	do {

		std::this_thread::sleep_for(sleepTimeMsec);

		processSampledData(getSamplerData());
		samplerState = getSamplerState();

	} while ((samplerState == nlc::SamplerState::Ready)
			 || (samplerState == nlc::SamplerState::Running));

	// Process any remaining data
	processSampledData(getSamplerData());

	if (samplerState == nlc::SamplerState::Failed)
		handleSamplerFailed();
}

void NanoLibSamplerExample::processSamplerWithoutNotificationRepetitive() {

	const std::chrono::milliseconds sleepTimeMsec(periodMilliseconds);

	std::cout << "\nSampler without notification in repetative mode:" << std::endl;

	configure(nlc::SamplerMode::Repetitive);
	start();

	nlc::SamplerState samplerState(nlc::SamplerState::Ready);
		
	do {

		std::this_thread::sleep_for(sleepTimeMsec);

		processSampledData(getSamplerData());

		if (lastIteration >= 4) {
			// In repetative mode the sampler will continue to run until it is stopped or an error occurs
			nanolibHelper.stopSampler(deviceHandle);
			break;
		}

		samplerState = getSamplerState();

	} while ((samplerState == nlc::SamplerState::Ready)
			 || (samplerState == nlc::SamplerState::Running));

	// Process any remaining data
	processSampledData(getSamplerData());

	if (samplerState == nlc::SamplerState::Failed)
		handleSamplerFailed();
}

void NanoLibSamplerExample::processSamplerWithoutNotificationContinuous() {

	const std::chrono::milliseconds sleepTimeMsec(periodMilliseconds);
	
	std::cout << "\nSampler without notification in continuous mode:" << std::endl;

	configure(nlc::SamplerMode::Continuous);
	start();

	nlc::SamplerState samplerState(nlc::SamplerState::Ready);
	constexpr unsigned maxCycles = 10;
	unsigned cycles = 0;

	do {

		std::this_thread::sleep_for(sleepTimeMsec);

		processSampledData(getSamplerData());

		if (++cycles == maxCycles) {
			// In continuous mode the sampler will continue to run until it is stopped or an error occurs
			nanolibHelper.stopSampler(deviceHandle);
			break;
		}

		samplerState = getSamplerState();

	} while ((samplerState == nlc::SamplerState::Ready)
			 || (samplerState == nlc::SamplerState::Running));

	// Process any remaining data
	processSampledData(getSamplerData());

	if (samplerState == nlc::SamplerState::Failed)
		handleSamplerFailed();
}

void NanoLibSamplerExample::processExamplesWithNotification() {

	processSamplerWithNotificationNormal();
	processSamplerWithNotificationRepetitive();
	processSamplerWithNotificationContinuous();
}

void NanoLibSamplerExample::processSamplerWithNotificationNormal() {

	const std::chrono::milliseconds sleepTimeMsec(periodMilliseconds);
		
	std::cout << "\nSampler with notification in normal mode:" << std::endl;

	configure(nlc::SamplerMode::Normal);

	SamplerNotifyExample samplerNotify(*this);

	start(&samplerNotify);
	while (samplerNotify.isRunning()) {
		std::this_thread::sleep_for(sleepTimeMsec);
	}
}

void NanoLibSamplerExample::processSamplerWithNotificationRepetitive() {

	const std::chrono::milliseconds sleepTimeMsec(periodMilliseconds);

	std::cout << "\nSampler with notification in repetative mode:" << std::endl;

	configure(nlc::SamplerMode::Repetitive);

	SamplerNotifyExample samplerNotify(*this);

	start(&samplerNotify);
	while (samplerNotify.isRunning()) {
		std::this_thread::sleep_for(sleepTimeMsec);

		if (lastIteration >= 4) {
			// In repetative mode the sampler will continue to run until it is stopped or an error occurs
			nanolibHelper.stopSampler(deviceHandle);
			break;
		}
	}
}

void NanoLibSamplerExample::processSamplerWithNotificationContinuous() {

	std::cout << "\nSampler with notification in continuous mode:" << std::endl;

	configure(nlc::SamplerMode::Continuous);

	SamplerNotifyExample samplerNotify(*this);

	start(&samplerNotify);
	std::this_thread::sleep_for(std::chrono::milliseconds(periodMilliseconds * 10));
	// In continuous the sampler will continue to run until it is stopped or an error occurs
	nanolibHelper.stopSampler(deviceHandle);
}

void NanoLibSamplerExample::configure(const nlc::SamplerMode mode) {

	constexpr size_t numberOfTrackedAddresses
		= sizeof(trackedAddresses) / sizeof(trackedAddresses[0]);

	nlc::SamplerConfiguration samplerConfiguration;

	for (size_t trackedAddressIndex = 0; trackedAddressIndex < numberOfTrackedAddresses;
		 ++trackedAddressIndex) {
		samplerConfiguration.trackedAddresses.push_back(
			trackedAddresses[trackedAddressIndex].odIndex);
	}

	samplerConfiguration.triggerAddress = triggerAddress;
	samplerConfiguration.triggerCondition = triggerCondition;
	samplerConfiguration.triggerValue = triggerValue;
	samplerConfiguration.periodMilliseconds = periodMilliseconds;
	samplerConfiguration.numberOfSamples = numberOfSamples;
	// Currrently this value is not used
	samplerConfiguration.preTriggerNumberOfSamples = 0;
	samplerConfiguration.mode = mode;
	samplerConfiguration.forceSoftwareImplementation = (mode == nlc::SamplerMode::Continuous);

	nanolibHelper.configureSampler(deviceHandle, samplerConfiguration);
}

void NanoLibSamplerExample::start(SamplerNotifyExample *samplerNotify /* = nullptr*/,
								  int64_t applicationData /* = 0*/) {

	lastIteration = std::numeric_limits<uint64_t>::max();
	sampleNumber = 0;
	headerPrinted = false;

	// Deactivate the start trigger
	nanolibHelper.writeInteger(deviceHandle, triggerValueInactive, triggerAddress, 32);

	// Start the sampler
	try {
		nanolibHelper.startSampler(deviceHandle, samplerNotify, applicationData);
	} catch (...) {
		if (samplerNotify != nullptr)
			samplerNotify->setInactive();
		throw;
	}

	// Activate start trigger
	try {
		nanolibHelper.writeInteger(deviceHandle, triggerValueActive, triggerAddress, 32);
	} catch (...) {
		nanolibHelper.stopSampler(deviceHandle);
		throw;
	}
}

nlc::SamplerState NanoLibSamplerExample::getSamplerState() {
	return nanolibHelper.getSamplerState(deviceHandle);
}

std::vector<nlc::SampleData> NanoLibSamplerExample::getSamplerData() {
	return nanolibHelper.getSamplerData(deviceHandle);
}

void NanoLibSamplerExample::handleSamplerFailed(
	const nlc::ResultVoid *lastErrorPtr /* = nullptr*/) {

	nlc::ResultVoid lastError;

	if (lastErrorPtr != nullptr) {
		lastError = *lastErrorPtr;
	} else {
		assert(getSamplerState() == nlc::SamplerState::Failed);
		lastError = nanolibHelper.getSamplerLastError(deviceHandle);
	}

	assert(lastError.hasError());
	std::cerr << std::endl
			  << "Sampler execution failed with error: " << lastError.getError() << std::endl;
}

void NanoLibSamplerExample::processSampledData(const std::vector<nlc::SampleData> &sampleDatas) {

	constexpr size_t numberOfTrackedAddresses
		= sizeof(trackedAddresses) / sizeof(trackedAddresses[0]);

	for (const auto sampleData : sampleDatas) {

		const auto &sampledValues = sampleData.sampledValues;
		const size_t numberOfSampledValues = sampledValues.size();

		assert((numberOfSampledValues % numberOfTrackedAddresses) == 0);

		if (lastIteration != sampleData.iterationNumber) {
			sampleNumber = 0;
			lastIteration = sampleData.iterationNumber;
		}
		
		std::stringstream stream;

		if (!headerPrinted) {

			static const char *cszHorzLine = "------------------------------------------------------------\n";

			stream << cszHorzLine;
			stream << std::left << std::setw(10) << "Iteration" << std::left << std::setw(10)
				   << "Sample";
			for (size_t trackedAddressIndex = 0; trackedAddressIndex < numberOfTrackedAddresses;
				 ++trackedAddressIndex) {

				std::string addressName("[");

				addressName += trackedAddresses[trackedAddressIndex].name;
				addressName += "]";
				stream << std::left << std::setw(14) << addressName << std::left << std::setw(8)
					   << "Time";
			}

			stream << "\n";
			stream << cszHorzLine;

			headerPrinted = true;
		}

		for (size_t index = 0; index < numberOfSampledValues; index += numberOfTrackedAddresses) {

			stream << std::left << std::setw(10) << lastIteration;
			stream << std::left << std::setw(10) << sampleNumber;

			for (size_t trackedAddressIndex = 0; trackedAddressIndex < numberOfTrackedAddresses;
				 ++trackedAddressIndex) {

				const auto &sampledValue = sampledValues[index + trackedAddressIndex];

				stream << std::left << std::setw(14) << sampledValue.value;
				stream << std::left << std::setw(8) << sampledValue.collectTimeMsec;
			}
			
			stream << "\n";
			++sampleNumber;
		}

		std::cout << stream.str();
	}
}